File: javascript/src/providers/TempmailPlus.js
import { TempMailProvider } from '../base.js';
import { Message, MessageDetail } from '../models.js';
import { TempMailError } from '../errors.js';

const BASE = 'https://tempmail.plus';
const DOMAIN = 'mailto.plus';
const CHARS = 'abcdefghijklmnopqrstuvwxyz0123456789';

/**
 * tempmail.plus provider — REST API, no auth, email passed as query param.
 */
export class TempmailPlus extends TempMailProvider {
  /** @param {{ randomUA?: boolean, useCookies?: boolean }|undefined} config */
  constructor(config = {}) {
    super('tempmail.plus', { randomUA: config.randomUA, useCookies: true });
    /** @type {string|null} */
    this._email = null;
  }

  /** @returns {Promise<string>} */
  async generateEmail() {
    const username = Array.from({ length: 10 }, () =>
      CHARS[Math.floor(Math.random() * CHARS.length)],
    ).join('');
    this._email = `${username}@${DOMAIN}`;
    return this._email;
  }

  /**
   * @param {string} email
   * @returns {Promise<Message[]>}
   */
  async getInbox(email) {
    const res = await this._fetch(`${BASE}/api/mails?email=${encodeURIComponent(email)}`);
    const data = await res.json();
    if (data.result === false) throw new TempMailError('API returned error', this.name);
    return (data.mail_list || []).map(item => new Message({
      id: String(item.mail_id ?? ''),
      sender: item.from_mail || '',
      subject: item.subject || '',
      date: item.time,
    }));
  }

  /**
   * @param {string} messageId
   * @returns {Promise<MessageDetail>}
   */
  async readMessage(messageId) {
    if (!this._email) {
      throw new TempMailError('No email — call generateEmail() first', this.name);
    }
    const url = `${BASE}/api/mails/${encodeURIComponent(messageId)}?email=${encodeURIComponent(this._email)}`;
    const res = await this._fetch(url);
    const data = await res.json();
    return new MessageDetail({
      id: String(data.mail_id ?? messageId),
      sender: data.from_mail || data.from || '',
      subject: data.subject || '',
      date: data.date,
      bodyText: data.text || '',
      bodyHtml: data.html || '',
      attachments: (data.attachments || []).map(a => ({
        filename: a.filename || '',
        contentType: a.content_type || '',
        url: '',
      })),
    });
  }

  /**
   * @param {string} _email
   * @returns {Promise<boolean>}
   */
  async deleteEmail(_email) {
    this._email = null;
    return true;
  }
}

File: rust/src/providers/tempmail_plus.rs
use async_trait::async_trait;
use serde::Deserialize;

use crate::error::TempMailError;
use crate::httpclient::HttpClient;
use crate::models::{Message, MessageDetail};
use crate::providers::TempMailProvider;

const BASE_URL: &str = "https://tempmail.plus";
const DOMAIN: &str = "mailto.plus";

/// Provider for tempmail.plus — REST API, no auth, email passed as query param.
#[derive(Debug)]
pub struct TempmailPlus {
    client: HttpClient,
    email: std::sync::Arc<tokio::sync::Mutex<Option<String>>>,
}

#[derive(Deserialize)]
struct InboxResp {
    #[serde(default)]
    result: Option<bool>,
    #[serde(default)]
    mail_list: Vec<serde_json::Value>,
}

impl TempmailPlus {
    pub fn new(client: HttpClient) -> Self {
        Self {
            client,
            email: std::sync::Arc::new(tokio::sync::Mutex::new(None)),
        }
    }
}

fn random_alphanumeric(len: usize) -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    const CHARSET: &[u8] = b"abcdefghijklmnopqrstuvwxyz0123456789";
    let mut s = String::with_capacity(len);
    let mut seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .subsec_nanos() as usize;
    for _ in 0..len {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) % (1 << 31);
        s.push(CHARSET[seed % CHARSET.len()] as char);
    }
    s
}

fn parse_date(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::NaiveDateTime::parse_from_str(s, "%Y-%m-%d %H:%M:%S")
        .map(|dt| dt.and_utc())
        .unwrap_or_else(|_| chrono::Utc::now())
}

fn parse_rfc2822(s: &str) -> chrono::DateTime<chrono::Utc> {
    chrono::DateTime::parse_from_rfc2822(s)
        .map(|dt| dt.with_timezone(&chrono::Utc))
        .unwrap_or_else(|_| chrono::Utc::now())
}

fn opt_str(v: Option<&serde_json::Value>) -> String {
    v.and_then(|x| x.as_str()).unwrap_or("").to_string()
}

fn opt_vid(v: Option<&serde_json::Value>) -> String {
    match v {
        Some(serde_json::Value::String(s)) => s.clone(),
        Some(serde_json::Value::Number(n)) => n.to_string(),
        Some(other) => other.to_string(),
        None => String::new(),
    }
}

#[async_trait]
impl TempMailProvider for TempmailPlus {
    async fn generate_email(&self) -> Result<String, TempMailError> {
        let username = random_alphanumeric(10);
        let email = format!("{}@{}", username, DOMAIN);
        let mut e = self.email.lock().await;
        *e = Some(email.clone());
        Ok(email)
    }

    async fn get_inbox(&self, email: &str) -> Result<Vec<Message>, TempMailError> {
        let resp: InboxResp = self
            .client
            .get_json(&format!("{}/api/mails?email={}", BASE_URL, email))
            .await?;
        if let Some(false) = resp.result {
            return Err(TempMailError::Api("tempmail.plus API returned error".into()));
        }
        Ok(resp
            .mail_list
            .into_iter()
            .map(|m| Message {
                id: opt_vid(m.get("mail_id")),
                sender: opt_str(m.get("from_mail")),
                subject: opt_str(m.get("subject")),
                date: parse_date(&opt_str(m.get("time"))), preview: String::new(), has_attachments: false })
            .collect())
    }

    async fn read_message(&self, message_id: &str) -> Result<MessageDetail, TempMailError> {
        let email = self
            .email
            .lock()
            .await
            .clone()
            .ok_or_else(|| TempMailError::Api("Must call generate_email first".into()))?;
        let m: serde_json::Value = self
            .client
            .get_json(&format!(
                "{}/api/mails/{}?email={}",
                BASE_URL, message_id, email
            ))
            .await?;
        Ok(MessageDetail {
            message: Message {
                id: opt_vid(m.get("mail_id")),
                sender: opt_str(m.get("from_mail").or_else(|| m.get("from"))),
                subject: opt_str(m.get("subject")),
                date: parse_rfc2822(&opt_str(m.get("date"))), preview: String::new(), has_attachments: false },
            body_text: {
                let b = opt_str(m.get("text"));
                if b.is_empty() { None } else { Some(b) }
            },
            body_html: {
                let h = opt_str(m.get("html"));
                if h.is_empty() { None } else { Some(h) }
            },
            attachments: vec![], ..Default::default()         })
    }

    async fn delete_email(&self, _email: &str) -> Result<bool, TempMailError> {
        let mut e = self.email.lock().await;
        *e = None;
        Ok(true)
    }
}
